src/mlir/consumer.cc

```diff
@@ -0,0 +1,762 @@
+    os << name.str();
+
+    return os.str();
+  }
+
+  std::tuple<size_t, Type, bool>
+  ASTConsumer::getField(Type type, llvm::StringRef fieldName)
+  {
+    auto structTy = type.dyn_cast<StructType>();
+    assert(structTy && "Bad type for field access");
+
+    auto name = structTy.getName();
+    auto& fieldMap = classFields[name];
+    size_t pos = 0;
+    for (auto f : fieldMap.fields)
+    {
+      // Found the field, return position
+      if (f == fieldName)
+      {
+        auto ty = fieldMap.types[pos];
+        return {pos, ty, true};
+      }
+
+      // Always recurse into all class fields
+      auto classField = fieldMap.types[pos].dyn_cast<StructType>();
+      if (classField)
+      {
+        auto [subPos, subTy, subFound] = getField(classField, fieldName);
+        pos += subPos;
+
+        if (subFound)
+          return {pos, subTy, true};
+        else
+          continue;
+      }
+
+      // Not found, continue searching
+      pos++;
+    }
+
+    // Not found, return false
+    return {0, Type(), false};
+  }
+
+  // ===================================================== AST -> MLIR
+  llvm::Error ASTConsumer::consumeRootModule(Ast ast)
+  {
+    auto node = nodeAs<Class>(ast);
+    assert(node && "Bad node");
+
+    // TODO: Quick early pass to declare all functions before we descend the
+    // AST.
+
+    // Modules are just global classes
+    return consumeClass(ast);
+  }
+
+  llvm::Error ASTConsumer::consumeClass(Ast ast)
+  {
+    auto node = nodeAs<Class>(ast);
+    assert(node && "Bad node");
+
+    StringRef modName = node->id.view();
+    // The root module has no name, doesn't need to be in the context
+    if (!modName.empty())
+      functionScope.push_back(modName);
+
+    // Push another scope for variables, functions and types
+    SymbolScopeT var_scope(symbolTable());
+    auto type = StructType::getIdentified(builder().getContext(), modName);
+
+    // Create an entry for the fields
+    classFields.emplace(modName, FieldOffset());
+
+    // Lower members, types, functions
+    for (auto sub : node->members)
+    {
+      switch (sub->kind())
+      {
+        case Kind::Class:
+        {
+          auto err = consumeClass(sub);
+          if (err)
+            return err;
+          functionScope.pop_back();
+          break;
+        }
+        case Kind::Using:
+          // Ignore for now as this is just a reference to the module name
+          // that will be lowered, but module names aren't being lowered
+          // now.
+          break;
+        case Kind::Function:
+        {
+          auto func = consumeFunction(sub);
+          if (auto err = func.takeError())
+            return err;
+          gen.push_back(*func);
+          break;
+        }
+        case Kind::Field:
+        {
+          auto res = consumeField(sub);
+          if (auto err = res.takeError())
+            return err;
+
+          // Update class-field map
+          auto field = nodeAs<Field>(sub);
```

Comment:
right now, yes. soon, we'll have two passes. The first, declares all types and functions. The second implements all functions. We don't have tests for using fields before declarations or methods calling functions that weren't yet declared precisely because of that. this is future work.